theano/sandbox/cuda/opt.py

```diff
@@ -445,6 +446,26 @@ def local_gpu_dot_to_dot22(node):
                                                 shape_out))]
     return False
 
+@local_optimizer(None)
+def local_assert_no_cpu_op(node):
+    if not isinstance(node.op, GpuOp) and all([var.owner and isinstance(var.owner.op,
+        HostFromGpu) for var in node.inputs]) and any([[c for c in var.clients
+                if isinstance(c[0].op, GpuFromHost)] for var in node.outputs]):
+            if config.assert_no_cpu_op == "warn":
+                _logger.warning(("CPU op %s is detected in the computational"
+                                 " graph") % node)
+            elif config.assert_no_cpu_op == "raise":
+                raise RuntimeError("The op %s is on CPU." % node)
+            elif config.assert_no_cpu_op == "pdb":
+                pdb.set_trace()
+
+    return None
+
+#Register the local_assert_no_cpu_op:
+assert_no_cpu_op = theano.tensor.opt.in2out(local_assert_no_cpu_op, name='assert_no_cpu_op')
+# 48.7 is after specialize device
+theano.compile.optdb.register('assert_no_cpu_op', assert_no_cpu_op, 48.7)
```

Comment:
There is a first instance of GPU specialization at 48.5 (https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/opt.py#L77), then it it applied again at 49.1 (https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/opt.py#L84)
I think the idea is that we can only transfer to the GPU elemwise operations with float32 as inputs and outputs, and sometimes this is the case after fusion but not before (when intermediate results are int for instance).